File: frontend/src/types/event.ts
/**
 * Event Types
 * 
 * Type definitions for events, ticket types, and event management.
 * 
 * @module EventTypes
 */

// Event Status Enum
export type EventStatus = 'draft' | 'published' | 'cancelled' | 'completed'

// Core Event Interface
export interface Event {
  id: string
  organizerId: string
  organizerName?: string
  title: string
  slug: string
  description: string | null
  venueName: string
  venueAddress: string | null
  city: string
  startDate: string
  endDate: string | null
  startTime: string
  endTime: string | null
  coverImageUrl: string | null
  status: EventStatus
  salesStartDate: string | null
  salesEndDate: string | null
  ticketStats: TicketStats
  createdAt: string
  updatedAt: string
}

// Ticket Statistics
export interface TicketStats {
  totalSold: number
  totalRevenue: number
  totalAttendees: number
  availableTickets: number
}

// Ticket Type Interface
export interface TicketType {
  id: string
  eventId: string
  name: string
  description: string | null
  price: number
  quantityAvailable: number
  quantitySold: number
  salesStart: string | null
  salesEnd: string | null
  isActive: boolean
  createdAt: string
  updatedAt: string
}

// Create Event Data
export interface CreateEventData {
  title: string
  description?: string
  venueName: string
  venueAddress?: string
  city: string
  startDate: string
  startTime: string
  endDate?: string
  endTime?: string
  coverImage?: File
  ticketTypes: Omit<TicketType, 'id' | 'quantitySold' | 'createdAt' | 'updatedAt'>[]
}

// Update Event Data
export interface UpdateEventData extends Partial<CreateEventData> {
  status?: EventStatus
}

// Event List Response
export interface EventListResponse {
  events: Event[]
  total: number
  page: number
  limit: number
  totalPages: number
}

// Event Analytics
export interface EventAnalytics {
  totalTickets: number
  totalRevenue: number
  checkinCount: number
  checkinPercentage: number
  dailySales: DailySalesData[]
  ticketBreakdown: TicketBreakdownData[]
  recentCheckins: RecentCheckinData[]
}

export interface DailySalesData {
  date: string
  tickets: number
  revenue: number
}

export interface TicketBreakdownData {
  name: string
  sold: number
  revenue: number
  percentage: number
}

export interface RecentCheckinData {
  attendeeName: string
  ticketType: string
  checkedInAt: string
}

// Event Status Configuration
export interface EventStatusConfig {
  label: string
  color: string
  bgColor: string
  borderColor: string
  icon: string
  description: string
  allowedActions: string[]
}

// Event Filter Options
export interface EventFilters {
  status?: EventStatus
  city?: string
  startDate?: string
  endDate?: string
  search?: string
}

File: backend/internal/models/event.go
package models

import (
	"github.com/google/uuid"
	"github.com/le-arch/EventHub-C5/internal/db/repo"
)

// CapacityRangeJSON is the JSON representation of int4range.
type CapacityRangeJSON struct {
    Lower  int32  `json:"lower"`
    Upper  int32  `json:"upper"`
    Bounds string `json:"bounds,omitempty"` // "[)", "(]", "[]", "()"
}


type CreateEventRequest struct {
	OrganizerID    uuid.UUID        `json:"organizer_id" `
	Title          string           `json:"title" binding:"required"`
	Slug           string           `json:"slug" binding:"required"`
	Description    string           `json:"description" binding:"required"`
	Venue          string           `json:"venue" binding:"required"`
	City           string           `json:"city" binding:"required"`
	StartDate      string           `json:"start_date" binding:"required"`
	EndDate        string           `json:"end_date"`
	StartTime      *string           `json:"start_time" binding:"required"`
	EndTime        *string           `json:"end_time" binding:"required"`
	CoverImageUrl  string           `json:"cover_image_url"`
	Status         repo.EventStatus          `json:"status"`
	SalesStartDate string           `json:"sales_start_date"`
	SalesEndDate   string           `json:"sales_end_date"`
	CapacityRange *CapacityRangeJSON `json:"capacity_range"`

}

type UpdateEventRequest struct {
    Title          *string `json:"title,omitempty"`
    Slug           *string `json:"slug,omitempty"`
    Description    string `json:"description,omitempty"`
    Venue          string `json:"venue,omitempty"`
    City           *string `json:"city,omitempty"`
    StartDate      *string `json:"start_date,omitempty"`  
    EndDate        *string `json:"end_date,omitempty"`
    StartTime      *string `json:"start_time,omitempty"`
    EndTime        *string `json:"end_time,omitempty"`
    CoverImageUrl  string `json:"cover_image_url,omitempty"`
    Status         *repo.EventStatus `json:"status,omitempty"`
    SalesStartDate *string `json:"sales_start_date,omitempty"`
    SalesEndDate   *string `json:"sales_end_date,omitempty"`
	CapacityRange *CapacityRangeJSON `json:"capacity_range,omitempty"`
}

type UpdateEventStatusRequest struct {
	Status   repo.EventStatus `json:"status" binding:"required"`
}

type UpdateEventStatus struct {
	Status   repo.EventStatus `json:"status" binding:"required,oneof=draft published cancelled archived"`
}

// for event tickets
type CreateTicketTypeRequest struct {
	Name             string `json:"name" binding:"required"`
    Description      string `json:"description"`
    Price            int32  `json:"price" binding:"required,min=0"`
    QuantityAvailable int32 `json:"quantity_available" binding:"required,min=0"`
    IsActive         *bool  `json:"is_active"`
}

type UpdateTicketTypeRequest struct {
    Name             *string `json:"name"`
    Description      string `json:"description"`
    Price            *int32  `json:"price"`
    QuantityAvailable *int32 `json:"quantity_available"`
    IsActive         *bool   `json:"is_active"`
}